### ISSUE TITLE
Support search new friends by phone number or user name.

### ISSUE THREAD

**Issue by su-chang** [wechaty/wechaty#1892]
**Is your feature request related to a problem? Please describe.**

Wechaty have `bot.Friendship.add()` method for add new friend, but we can not make sure the new friend is who we want to add, so we hope Wechaty could support search friend, and then we could check the new friend by this API.

**Describe the solution you'd like**

### Base

In primary level, the `add()` method has two steps :

- search friends and get the `v1` and `v2` attributes for add friends
- add friends by `v1` and `v2` (_`v1` changeless, `v2` has expire time_)

### Design

So let's make `search()` as a static method for `Friendship`.

#### wechaty

```js
public static search (contact: Contact): Promise<Contact>
```

#### wechaty-puppet

```js
public abstract friendshipSearch (contactId: string): Promise<void>
```

#### wechaty-puppet-padplus

In `friendshipSearch()` method, there has two important things to do:

- save `v1`, `v2` data to cache
- abstract data to generate one contact and save it to cache as not friend

There's no need to request `v1`, `v2` again if they are in cache, so we could get them from cache and use them in `add()` method.

### Example
```js
const _contact = bot.Contact.load('phone number')
const contact: Contact = await bot.Friendship.search(_contact)

if (contact.name() === 'nick name') {
  await bot.Friendship.add(contact, 'hello')
}
```

### Question
1. For the type of param in `search()`, I think `string` is better than `Contact`, easy to use indeed.

2. After we searched new friend, it will generate one `Contact` which is not friend, so we should delete this `Contact` when `add()` method finished. Or we should save it for 24 hours (_maybe it due to the expire time of `v2`_) for one case : 

    > If the friend has not accept friend request, and the bot want to add again.

**Describe alternatives you've considered**

**Additional context**

[enhancement]

**Comment by huan** [comment]
> we can not make sure the new friend is who we want to add

Cloud you please explain that what's the meaning of the above description?

In other words, what's the purpose of this new API?

**Comment by su-chang** [comment]
If we want to add new friend, the first step we need to search him and then the next step add the new friend.

The purpose is for the first step.

**Comment by huan** [comment]
Thanks for your quick response.

If we want to search for anyone by their mobile phone number, I would like to use the following API:

```ts
const contact = await bot.Friendship.search({ phone: '13913813711' })
if (contact) {
  await bot.Friendship.add(contact)
}
```

I feel the above code will be easy to read and be understood, please feel free to let me know what you think.

**Comment by su-chang** [comment]
How about the user name? We know we can search friends by two ways : phone and user name.

I think it's better to use string parameter for `search`.

```js
const contact = await bot.Friendship.search(phoneOrUserName)
if (contact) {
  await bot.Friendship.add(contact)
}
```

**Comment by huan** [comment]
We can support the weixin Id by defining:

```
{
  phone?: string,
  weixin?: string,
}
```

Then we can support two ways for the search.

**Comment by windmemory** [comment]
```ts
{
  phone?: string,
  weixin?: string,
}
```
This design should solve the problem.

**Comment by huan** [comment]
Implemented by #1895
